File: frameworks/node/hono/Implementation/routes/query.ts
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";

import type { Routes } from "../app.ts";

// rb:wiring query.*
/**
 * The query string as each route binds it. zValidator's query target hands zod the parsed query,
 * whose values are strings, and zod converts the numbers to integers, so c.req.valid is the echo.
 */
const integer = z.coerce.number().int();

const page = z.object({ page: integer });

/** query.many's eight values, which forms.urlencoded posts as a form. */
export const search = z.object({
  page: integer,
  size: integer,
  status: z.string(),
  category: z.string(),
  sort: z.string(),
  q: z.string(),
  minPrice: integer,
  maxPrice: integer,
});
// rb:end

/** query: query string values bound by zValidator and echoed back. */
const query: Routes = (app, p) => {
  app.get("/query/one", zValidator("query", page), (c) => c.json({ ...p.small, echo: c.req.valid("query") }));

  app.get("/query/many", zValidator("query", search), (c) => c.json({ ...p.small, echo: c.req.valid("query") }));
};

export default query;

File: frameworks/rust/rocket/UnitTests/query.rs
use serde_json::json as literal;

use crate::support::{client, get, json, with_echo};

// rb:test query.one
/// query.one: one value bound as an integer and echoed.
#[test]
fn one_value_is_bound() {
    let client = client();

    let response = get(&client, "/query/one?page=417", &[]);

    assert_eq!(json(response), with_echo("items.small.json", literal!({ "page": 417 })));
}

// rb:test query.many
/// query.many: eight values bound, percent-decoded and echoed, the numbers as integers.
#[test]
fn eight_values_are_bound() {
    let client = client();
    let path = "/query/many?page=417&size=38&status=paid&category=garden&sort=created&q=alpha%20bravo&minPrice=1200&maxPrice=34000";

    let response = get(&client, path, &[]);

    let echo = literal!({ "page": 417, "size": 38, "status": "paid", "category": "garden", "sort": "created", "q": "alpha bravo", "minPrice": 1200, "maxPrice": 34000 });
    assert_eq!(json(response), with_echo("items.small.json", echo));
}
